accepted/00000-maintenance-windows.md

```diff
@@ -0,0 +1,99 @@
+- Feature Name: Maintenance Windows
+- Start Date: 2020-03-31
+- RFC PR: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Define Maintenance Windows in Uyuni and assign them to systems
+
+# Motivation
+[motivation]: #motivation
+
+Customers have company policies to perform changes on systems only at specific points in time.
+This is typically called "Maintenance Window".
+
+Uyuni should be able to define them and prevent executing of actions outside of such a window.
+
+# Architecture
+[architecture]: #architecture
+
+A system is in "Maintenance Mode" or not.
+When a system is in "Maintenance Mode" no action can be executed.
+Setting the Maintenance Mode "on" and "off" requires Organization Admin permissions.
+A "Maintenance Window" does nothing else than turning Maintenance Mode "on" when the window starts
+and "off" when it ends.
+
+Defining Maintenance Windows requires Organization Admin permissions.
+
+Maintenance Windows are defined on system level.
+Every System can get 1 Schedule assigned.
+A Schedule can be used for multiple systems.
+Inside of the Schedule, multiple Maintenance Windows can be defined.
+They can be defined as individual or as recurring events.
+
+To assign a Schedule to multiple systems, SSM should be used.
+
+
+# Detailed design
+[design]: #detailed-design
+
+## Implement a Maintenance Mode
+
+Check, how System Locking is implemented. Copy the mechanism for Maintenance Mode.
+Does it make sense to use System Locking as "Maintenance Mode"?
+
+For the first iteration we want to block all actions when the system is not in maintenance mode.
+In future iterations we can consider to allow some actions which are mostly read-only (package list refresh or hardware refresh).
+Also actions which prepare things for the maintenance window could be considered (pre-fetching packages for update).
+
+
+## Maintenance Schedule
+
+A Maintenance Schedule is a set of Maintenance Windows. A Maintenance Schedule has a name/label.
+The data should be stored as ICal data (https://tools.ietf.org/html/rfc5545).
```

Comment:
I wonder if we want to store the iCal data as a varchar in the db. If we do it this way, we need to be prepared that some queries, we might need, can be costly (in general, it would mean selecting everything to memory and do the filtering on backend).

Reply:
Or we convert ical data as single events into a DB structure. This requires of cause update capability. On the other hand it would make it easier to get these data also in python and other languages when needed.

Reply:
If we convert the iCal into single events how would we do the updating? I guess we don't want to override the entire table every time. (Could do a check what changed on frontend. What about xmlrpc?)

Reply:
It should be possible to create the schedule via the API

Reply:
Added some notes about XMLRPC API usage to the RFC.